src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq_vmr.c

```diff
@@ -3019,43 +3021,109 @@ static ssize_t program_vmr_store(struct device *dev,
+        ret = xgq_vmr_default_fpt(xgq, &log_buf, &log_size, off, count);
+        if (ret)
+		return -EINVAL;
 
-	if (ret)
-		return ret;
+        if (off >= log_size)
+                goto done;
 
-	mutex_lock(&xgq->xgq_lock);
+        if (off + count > log_size)
+                count = log_size - off;
 
-	blob = xgq->xgq_vmr_system_dtb;
-	size = xgq->xgq_vmr_system_dtb_size;
+        memcpy(buf, log_buf + off, count);
+        ret = count;
+done:
+	vfree(log_buf);
+        XGQ_INFO(xgq,"Offset  = %d, count = %d, ret = %d\n",(int)off,(int)count,(int)ret);
+        return ret;
+}
+
+static struct bin_attribute bin_attr_vmr_default_fpt = {
+        .attr = {
+                .name = "vmr_default_fpt",
+                .mode = 0444
+        },
+        .read = vmr_default_fpt_read,
+        .write = NULL,
+        .size = 0
+};
 
-	if (off >= size)
-		goto out;
+static ssize_t vmr_backup_fpt_read(struct file *filp, struct kobject *kobj,
+        struct bin_attribute *attr, char *buf, loff_t off, size_t count)
+{
+        struct xocl_xgq_vmr *xgq =
+                dev_get_drvdata(container_of(kobj, struct device, kobj));
+        char *log_buf = NULL;
+        size_t log_size = 0;
+        ssize_t ret = 0;
 
-	if (off + count > size)
-		count = size - off;
-	memcpy(buf, blob + off, count);
+        /* return count should be less or equal to count */
+        ret = xgq_vmr_backup_fpt(xgq, &log_buf, &log_size, off, count);
+        if (ret)
+		return -EINVAL;
 
-	ret = count;
-out:
-	mutex_unlock(&xgq->xgq_lock);
-	return ret;
-}
+        if (off >= log_size)
+                goto done;
+
+        if (off + count > log_size)
+                count = log_size - off;
 
+        memcpy(buf, log_buf + off, count);
+        ret = count;
+done:
+	vfree(log_buf);
+        XGQ_INFO(xgq,"Offset  = %d, count = %d, ret = %d\n",(int)off,(int)count,(int)ret);
+        return ret;
+}
+
+static struct bin_attribute bin_attr_vmr_backup_fpt = {
+        .attr = {
+                .name = "vmr_backup_fpt",
+                .mode = 0444
+        },
+        .read = vmr_backup_fpt_read,
+        .write = NULL,
+        .size = 0
+};
+
+static ssize_t vmr_system_dtb_read(struct file *filp, struct kobject *kobj,
+        struct bin_attribute *attr, char *buf, loff_t off, size_t count)
+{
+        struct xocl_xgq_vmr *xgq =
+                dev_get_drvdata(container_of(kobj, struct device, kobj));
+        char *log_buf = NULL;
+        size_t log_size = 0;
+        ssize_t ret = 0;
+
+        /* return count should be less or equal to count */
+        ret = xgq_vmr_system_dtb(xgq, &log_buf, &log_size, off, count);
+        if (ret)
+		return -EINVAL;
+
+        if (off >= log_size)
+                goto done;
+
+        if (off + count > log_size)
+                count = log_size - off;
+
+        memcpy(buf, log_buf + off, count);
+        ret = count;
+done:
+	vfree(log_buf);
+        XGQ_INFO(xgq,"Offset  = %d, count = %d, ret = %d\n",(int)off,(int)count,(int)ret);
```

Comment:
nits: to avoid compiling warning, the more cleaner way: offset and count is %ld, ret is %d